File: credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

int luhn(int card[], int n); // Luhn Algorithm

int main(void)
{
    long number, num, x; // number credit card
    int n = 1; // counter number of digits
    int *card; // array for number of digits
          // get number credit card
    number = get_long("Number: ");
    while (number < 0)
    {
        number = get_long("Number: ");
    }
    num = number;
        // counter the number of digits
    while ((num/=10) > 0)
    {
        n++;
    }
        // переставляємо цифри карти задом наперед для зручності
    num = number;
    x = 0;
    while (num > 0)
    {
        x = x * 10 + (num%10);
        num /= 10;
    }
        // виділяємо пам'ять під масив для номера карти для зручності обрахунку
    card = malloc(n * sizeof(int));
    num = x;
    if (n == 13 || n == 15 || n == 16)
    {
            // заносимо номер карти в масив
        for (int i = 0; i < n; i++)
        {
            card[i] = num%10;
            num /= 10;
        }
            // check credit card
        if (luhn(card, n) == 0)
        {
            if (card[0] == 4)
            {
                printf("VISA\n");
            }
            else if (card[0] == 3)
            {
                if (card[1] == 4 || card[1] == 7)
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else if (card[0] == 5)
            {
                if (card[1] == 1 || card[1] == 2 || card[1] == 3 || card[1] == 4 || card[1] == 5)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else
            {
                 printf ("INVALID\n");
            }
        }
        else
        {
            printf ("INVALID\n");
        }

    }
    else
    {
        printf ("INVALID\n");
    }
    free (card);
}

int luhn(int card[], int n)
{
    int l;
    int sum1 = 0;
    int sum2 = 0;
    if ((n%10)/2 == 3)
    {
        for (int i = 0; i < n + 1; i +=2)
        {
            if (card[i]*2 > 9)
            {
                sum1 += ((card[i]*2)%10 + (card[i]*2)/10);
            }
            else
            {
                sum1 += card[i]*2;
            }
        }
        for (int j = 1; j < n + 1; j +=2)
        {
            sum2 += card[j];
        }
    }
    else
    {
        for (int i = 1; i < n + 1; i +=2)
        {
            if (card[i]*2 > 9)
            {
                sum1 += ((card[i]*2)%10 + (card[i]*2)/10);
            }
            else
            {
                sum1 += card[i]*2;
            }
        }
        for (int j = 0; j < n + 1; j +=2)
        {
            sum2 += card[j];
        }
    }

    l = (sum1 + sum2)%10;
    return l;
}

File: credit.py
while True:
    n = input("Number: ")
    if n.isdigit():
        break
strlen = len(n)

# алгоритм Luhn (Validation of a credit card number)
product_sum = 0
if strlen % 2 == 0:
    amount = int(strlen / 2)
else:
    amount = strlen // 2
x = strlen
for i in range(amount):
    if 2 * int(n[x - 2]) > 9:
        product_sum += (2 * int(n[x - 2])) % 10 + 1
        x -= 2
    else:
        product_sum += 2 * int(n[x - 2])
        x -= 2
x = strlen + 1
sum_2 = 0
for i in range(strlen - amount):
    sum_2 += int(n[x - 2])
    x -= 2
if (product_sum + sum_2) % 10 != 0:
    print("INVALID")

# перевірка
elif strlen == 13 or strlen == 15 or strlen == 16:
    if strlen == 13 or strlen == 16:
        if int(n[0]) == 4:
            print("VISA")
        elif int(n[0:2]) == 51 or int(n[0:2]) == 52 or int(n[0:2]) == 53 or int(n[0:2]) == 54 or int(n[0:2]) == 55:
            print("MASTERCARD")
        else:
            print("INVALID")
    elif strlen == 15:
        if int(n[0:2]) == 34 or int(n[0:2]) == 37:
            print("AMEX")
        else:
            print("INVALID")
    else:
        print("INVALID")
else:
    print("INVALID")
